src/main/kotlin/org/rust/ide/inspections/RsWrongTypeParametersNumberInspection.kt

```diff
@@ -16,18 +17,51 @@ class RsWrongTypeParametersNumberInspection : RsLocalInspectionTool() {
                 // Don't apply generic declaration checks to Fn-traits and `Self`
                 if (type.path?.valueParameterList != null) return
                 if (type.path?.cself != null) return
-
-                val paramsDecl = type.path?.reference?.resolve() as? RsGenericDeclaration ?: return
-                val expectedRequiredParams = paramsDecl.typeParameterList?.typeParameterList?.filter { it.eq == null }?.size ?: 0
-                val expectedTotalParams = paramsDecl.typeParameterList?.typeParameterList?.size ?: 0
-                val actualArgs = type.path?.typeArgumentList?.typeReferenceList?.size ?: 0
-                val (code, expectedText) = when {
-                    actualArgs < expectedRequiredParams -> ("E0243" to if (expectedRequiredParams != expectedTotalParams) "at least $expectedRequiredParams" else "$expectedTotalParams")
-                    actualArgs > expectedTotalParams -> ("E0244" to if (expectedRequiredParams != expectedTotalParams) "at most $expectedTotalParams" else "$expectedTotalParams")
-                    else -> null
-                } ?: return
-                holder.registerProblem(type, "Wrong number of type parameters: expected $expectedText, found $actualArgs [$code]")
+                checkMethod(holder, type)
             }
+
+            override fun visitCallExpr(o: RsCallExpr) = checkMethod(holder, o)
+            override fun visitMethodCallExpr(o: RsMethodCallExpr) = checkMethod(holder, o)
+        }
+
+    private fun checkMethod(holder: ProblemsHolder, o: RsCompositeElement) {
+        val actualArgs = when (o) {
+            is RsMethodCallExpr -> o.typeArgumentList?.typeReferenceList?.size ?: 0
+            is RsCallExpr -> (o.expr as RsPathExpr?)?.path?.typeArgumentList?.typeReferenceList?.size ?: 0
```

Comment:
Did you mean `o.expr as? RsPathExpr`? Otherwise this can produce ClassCastExceptions.